x-pack/plugins/cases/server/client/cases/update_custom_field.ts

```diff
@@ -0,0 +1,177 @@
+import { flattenCaseSavedObject } from '../../common/utils';
+import { decodeOrThrow } from '../../../common/api/runtime_types';
+import type { Case } from '../../../common/types/domain';
+import { CaseRt } from '../../../common/types/domain';
+import { decodeWithExcessOrThrow } from '../../../common/api';
+import {
+  validateCustomFieldKeysAgainstConfiguration,
+  validateCustomFieldTypesInRequest,
+} from './validators';
+import type { UserActionEvent } from '../../services/user_actions/types';
+
+export interface UpdateCustomFieldArgs {
+  /**
+   * The ID of a case
+   */
+  caseId: string;
+  /**
+   * The ID of a custom field to be updated
+   */
+  customFieldId: string;
+  /**
+   * value of custom field to update, case version
+   */
+  request: CustomFieldPatchRequest;
+}
+
+/**
+ * Updates the specified cases with new values
+ *
+ * @ignore
+ */
+export const updateCustomField = async (
+  { caseId, customFieldId, request }: UpdateCustomFieldArgs,
+  clientArgs: CasesClientArgs,
+  casesClient: CasesClient
+): Promise<Case> => {
+  const {
+    services: { caseService, userActionService },
+    user,
+    logger,
+    authorization,
+  } = clientArgs;
+
+  try {
+    const { value, caseVersion } = request;
+
+    const caseToUpdate = await caseService.getCase({
+      id: caseId,
+    });
+
+    const configurations = await casesClient.configure.get({
+      owner: caseToUpdate.attributes.owner,
+    });
+
+    await authorization.ensureAuthorized({
+      entities: [{ owner: caseToUpdate.attributes.owner, id: caseToUpdate.id }],
+      operation: Operations.updateCase,
+    });
+
+    const foundCustomField = configurations[0]?.customFields.find(
+      (item) => item.key === customFieldId
+    );
+
+    if (!foundCustomField) {
+      throw Boom.badRequest('cannot find custom field');
+    }
+
+    const decodedCustomFields = decodeWithExcessOrThrow(CustomFieldsRt)([
+      {
+        value,
+        type: foundCustomField.type,
+        key: customFieldId,
+      },
+      ...caseToUpdate.attributes.customFields.filter((field) => field.key !== customFieldId),
+    ]);
+
+    const updatedAt = new Date().toISOString();
+
+    if (value == null && foundCustomField.required) {
+      throw Boom.badRequest('Custom field value cannot be null or undefined.');
+    }
+
+    validateCustomFieldKeysAgainstConfiguration({
+      requestCustomFields: decodedCustomFields,
+      customFieldsConfiguration: configurations[0].customFields,
+    });
+
+    validateCustomFieldTypesInRequest({
+      requestCustomFields: decodedCustomFields,
+      customFieldsConfiguration: configurations[0].customFields,
+    });
+
+    const customFieldsToUpdate: CaseRequestCustomFields = caseToUpdate.attributes.customFields.map(
+      (cf) => {
+        if (cf.key === customFieldId) {
+          return {
+            value,
+            type: foundCustomField.type,
+            key: customFieldId,
+          } as CaseRequestCustomField;
```

Comment:
Added this because TS gives error that `toggle custom field type` is not compatible with `text custom field type`.

Reply:
Every single time I have this problem 😆

Reply:
nit: It seems that you have the same code above in L87. What about?

```
const customFieldsToUpdate = {
        value,
        type: foundCustomField.type,
        key: customFieldId,
      },
      ...caseToUpdate.attributes.customFields.filter((field) => field.key !== customFieldId)
 }
 
 const decodedCustomFields = decodeWithExcessOrThrow(CustomFieldsRt)(customFieldsToUpdate)
 
 const patchCasesPayload = {
      caseId,
      originalCase: caseToUpdate,
      updatedAttributes: {
        customFields: decodedCustomFields,
        updated_at: updatedAt,
        updated_by: user,
      },
      version: caseVersion,
    };
```